### PR TITLE
Aidan Mo: Completed Onboarding Challenge

### PR DESCRIPTION
**Pull request by Aloe-Verified**

### CODE REVIEW

File: lm75bd/lm75bd.c [hunk 1]
```diff
@@ -27,7 +27,26 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
 
 error_code_t readTempLM75BD(uint8_t devAddr, float *temp) {
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c:28]: Check if temp is not null before use, returning invalid arg

File: lm75bd/lm75bd.c [hunk 2]
```diff
@@ -27,7 +27,26 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
...
   /* Implement this driver function */
+  uint8_t buffer[1];
+  buffer[0] = 0x00;
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]:
```suggestion
uint8_t buffer[1] ={0};
```

File: lm75bd/lm75bd.c [hunk 3]
```diff
@@ -27,7 +27,26 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
...
+  error_code_t error = i2cSendTo(devAddr, buffer, 1);
+    if (error != ERR_CODE_SUCCESS) {
+        return error;
+    }
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]: Use the RETURN_IF_ERROR_CODE macro instead

File: lm75bd/lm75bd.c [hunk 4]
```diff
@@ -27,7 +27,26 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
 
 error_code_t readTempLM75BD(uint8_t devAddr, float *temp) {
   /* Implement this driver function */
+  uint8_t buffer[1];
+  buffer[0] = 0x00;
+  error_code_t error = i2cSendTo(devAddr, buffer, 1);
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]: Use the sizeof(buffer) instead of 1

File: services/thermal_mgr/thermal_mgr.c [hunk 5]
```diff
@@ -41,20 +41,49 @@ void initThermalSystemManager(lm75bd_config_t *config) {
 
 }
 
+#define ERR_CODE_FAIL 1
 error_code_t thermalMgrSendEvent(thermal_mgr_event_t *event) {
   /* Send an event to the thermal manager queue */
-
-  return ERR_CODE_SUCCESS;
+  if(xQueueSend(thermalMgrQueueHandle, event, portMAX_DELAY) == pdTRUE){
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c]: Check that the event and queue handles aren't null, returning invalid arg and invalid state respectively

File: lm75bd/lm75bd.c [hunk 6]
```diff
@@ -27,7 +27,26 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
...
+    if (error != ERR_CODE_SUCCESS) {
+        return error;
+    } 
   
+  uint8_t temperatureData[2];
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]: Initialize buffer

File: lm75bd/lm75bd.c [hunk 7]
```diff
@@ -27,7 +27,26 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
...
+
+  error = i2cReceiveFrom(devAddr, temperatureData, 2);
+    if (error != ERR_CODE_SUCCESS) {
+      return error;
+    }
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]: Use RETURN_IF_ERROR_CODE macro

File: lm75bd/lm75bd.c [hunk 8]
```diff
@@ -27,7 +27,26 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
...
+  uint8_t buffer[1];
+  buffer[0] = 0x00;
+  error_code_t error = i2cSendTo(devAddr, buffer, 1);
+    if (error != ERR_CODE_SUCCESS) {
+        return error;
+    } 
   
+  uint8_t temperatureData[2];
+
+  error = i2cReceiveFrom(devAddr, temperatureData, 2);
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]: Use sizeof(temperatureData) instead of 2

File: services/thermal_mgr/thermal_mgr.c [hunk 9]
```diff
@@ -41,20 +41,49 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
+    return ERR_CODE_SUCCESS;
+  }
+  return ERR_CODE_FAIL;
 }
-
+#define THERMAL_MGR_EVENT_OS_INTERRUPT 1
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c]: This should be added to the enum in the thermal mgr header file

File: services/thermal_mgr/thermal_mgr.c [hunk 10]
```diff
@@ -41,20 +41,49 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
+  thermalMgrSendEvent(&event);
+
 }
 
 static void thermalMgr(void *pvParameters) {
   /* Implement this task */
+  lm75bd_config_t *config = (lm75bd_config_t *) pvParameters;
+  uint8_t thermalSensorAddr = config->devAddr;
+  thermal_mgr_event_t event;
+  float temp;
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c]: Move the scope of temp down

File: services/thermal_mgr/thermal_mgr.c [hunk 11]
```diff
@@ -41,20 +41,49 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
+
+
   while (1) {
-    
+    if(xQueueReceive(thermalMgrQueueHandle, &event, portMAX_DELAY) == pdTRUE) {
+      if (event.type == THERMAL_MGR_EVENT_MEASURE_TEMP_CMD) {
+        if(readTempLM75BD(thermalSensorAddr, &temp) == ERR_CODE_SUCCESS){
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c]: We want to be logging that a read to the temperature sensor failed in both cases

File: services/thermal_mgr/thermal_mgr.c [hunk 12]
```diff
@@ -41,20 +41,49 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
+          addTemperatureTelemetry(temp);
+        }
+      }
+      else if (event.type == THERMAL_MGR_EVENT_OS_INTERRUPT){
+        if(readTempLM75BD(thermalSensorAddr, &temp) == ERR_CODE_SUCCESS){
+          if(temp >= 80.0f){
+            overTemperatureDetected();
+          }
+          else if (temp < 75.0f){
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c]: Use the macros defined in the thermal manager header file instead of hardcoding these values

**Review by Yarik-Popov — CHANGES_REQUESTED** [review]
Overall good start. Some changes are needed

File: lm75bd/lm75bd.c [hunk 13]
```diff
@@ -27,7 +27,25 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
...
+
+  uint8_t MSB = temperatureData[0];
+  uint8_t LSB = temperatureData[1];
+  int16_t bit_temperature = ((MSB << 8) | LSB) >> 5;
+  if (bit_temperature & (1 << 10)) { 
+    bit_temperature |= 0xF800;
+  }
+  if (temp == NULL) {
+    return ERR_CODE_INVALID_ARG;
+  }
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]: This should be at the top. No point of calling driver functions or doing computations if you can't store it in temp which is a simple checl

**Review by Yarik-Popov — CHANGES_REQUESTED** [review]
One minor change left

File: services/thermal_mgr/thermal_mgr.c [hunk 14]
```diff
@@ -43,21 +44,57 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
   /* Send an event to the thermal manager queue */
-
-  return ERR_CODE_SUCCESS;
+  if (event == NULL){
+    return ERR_CODE_INVALID_ARG;
+  }
+  if (thermalMgrQueueHandle == NULL) {
+    return ERR_CODE_INVALID_STATE;
+  }
+  if(xQueueSend(thermalMgrQueueHandle, event, portMAX_DELAY) == pdTRUE){
```
Comment (kepler452b123) [inline services/thermal_mgr/thermal_mgr.c]: this is too long to wait for queue to become lower- consider shortening it to a more reasonable value

**Review by kepler452b123 — CHANGES_REQUESTED** [review]
looks good, just one change

**Review by Yarik-Popov — COMMENTED** [review]
Lgtm, dm me your email and I'll add you to the notion

**Pull request closed without merging**